framework/src/partitioner/GridPartitioner.C

```diff
@@ -66,22 +75,65 @@ GridPartitioner::_do_partition(MeshBase & mesh, const unsigned int /*n*/)
   const auto & max = bounding_box.max();
 
   auto dim = mesh.mesh_dimension();
-  //  Need to make sure the number of cells in the grid matches the number of procs to partition for
-  nx = getParam<unsigned int>("nx");
 
+  // Gather user parameters
+  nx = isParamValid("nx") ? getParam<unsigned int>("nx") : nx;
   if (dim >= 2)
-    ny = getParam<unsigned int>("ny");
-
+    ny = isParamValid("ny") ? getParam<unsigned int>("ny") : ny;
   if (dim == 3)
-    nz = getParam<unsigned int>("nz");
+    nz = isParamValid("nz") ? getParam<unsigned int>("nz") : nz;
+
+  // simple info message unused parameters as this can be normal: we could be partitioning
+  // a 2D mesh before extruding it to 3D. The nz parameter is needed for the 3D mesh
+  if (dim < 2 && isParamValid("ny") && getParam<unsigned int>("ny") > 1)
+    paramInfo("ny", "Parameter ignored as mesh is currently of dimension less than 2.");
+  if (dim < 3 && isParamValid("nz") && getParam<unsigned int>("nz") > 1)
+    paramInfo("nz", "Parameter ignored as mesh is currently of dimension less than 3.");
+
+  // User parameters, which should match the number of partitions needed
+  if (getParam<MooseEnum>("grid_computation") == "manual")
+  {
+    //  Need to make sure the number of grid cells matches the number of procs to partition for
+    if ((nx * ny * nz) != mesh.n_partitions())
+      mooseError("Number of grid cells (" + std::to_string(nx * ny * nz) +
+                 ") does not match the number of MPI processes (" +
+                 std::to_string(mesh.n_partitions()) + ")");
+  }
 
-  // We should compute a balanced factorization so
-  // that we can assign proper processors to each direction.
-  // I just want to make grid partitioner smarter.
-  if ((nx * ny * nz) != mesh.n_partitions())
+  else if (getParam<MooseEnum>("grid_computation") == "automatic")
   {
-    // Anybody knows we are living in a 3D space.
-    int dims[] = {0, 0, 0};
+    // remove over-constraint and tell user
+    if (nx * ny * nz > mesh.n_partitions())
+    {
+      nx = 0;
+      ny = 0;
+      nz = 0;
+      paramWarning("grid_computation",
+                   "User specified (nx,ny,nz) grid exceeded number of partitions, these parameters "
+                   "will be ignored.");
+    }
+    // 0 means no restriction on which number to choose
+    std::array<int, 3> dims = {isParamValid("nx") ? int(nx) : 0,
+                               isParamValid("ny") ? int(ny) : 0,
+                               isParamValid("nz") ? int(nz) : 0};
+
```

Comment:
this is failing with
```
/tmp/build/moose/framework/src/partitioner/GridPartitioner.C: In member function 'virtual void GridPartitioner::_do_partition(libMesh::MeshBase&, unsigned int)':
/tmp/build/moose/framework/src/partitioner/GridPartitioner.C:137:47: error: cannot convert 'std::array<int, 3>' to 'int*'
  137 |     MPI_Dims_create(mesh.n_partitions(), dim, dims);
      |                                               ^~~~
      |                                               |
      |                                               std::array<int, 3>
In file included from /opt/libmesh/include/libmesh/libmesh_common.h:61,
```
I ll go back to the classic int []

Reply:
Nah, use https://en.cppreference.com/w/cpp/container/array/data instead

Reply:
Ok I ll get that in a future PR